crates/ide-assists/src/handlers/generate_delegate_trait.rs

```diff
@@ -0,0 +1,1064 @@
+                name = l.fields().position(|it| it == f)?.to_string();
+                ty = f.ty()?;
+                range = f.syntax().text_range();
+            }
+        };
+
+        let hir_ty = ctx.sema.resolve_type(&ty)?;
+        let type_impls = hir::Impl::all_for_type(db, hir_ty.clone());
+        let mut impls = Vec::with_capacity(type_impls.len());
+        let type_param = hir_ty.as_type_param(db);
+
+        if let Some(tp) = type_param {
+            for tb in tp.trait_bounds(db) {
+                impls.push(Delegee::Bound(BoundCase { 0: tb }));
+            }
+        };
+
+        for imp in type_impls {
+            match imp.trait_(db) {
+                Some(tr) => {
+                    if tr.is_visible_from(db, module) {
+                        impls.push(Delegee::Impls(ImplCase { 0: tr, 1: imp }))
+                    }
+                }
+                None => {
+                    continue;
+                }
+            }
+        }
+
+        Some(Field { name, ty, range, impls })
+    }
+}
+
+/// A field that we want to delegate can offer the enclosing struct
+/// trait to implement in two ways. The first way is when the field
+/// actually implements the trait and the second way is when the field
+/// has a bound type parameter. We handle these cases in different ways
+/// hence the enum.
+enum Delegee {
+    Bound(BoundCase),
+    Impls(ImplCase),
+}
+
+struct BoundCase(hir::Trait);
+struct ImplCase(hir::Trait, hir::Impl);
+
+/// When we list traits we can implement for the enclosing struct
+/// we use the absolute path of a trait. This trait consists of a single
+/// method that produces this path.
+trait Signatured {
+    fn signature(&self, db: &dyn HirDatabase) -> String;
+}
+
+impl Signatured for Delegee {
+    fn signature(&self, db: &dyn HirDatabase) -> String {
+        let mut s = String::new();
+        let trait_: hir::Trait;
+
+        match self {
+            Delegee::Bound(b) => trait_ = b.0,
+            Delegee::Impls(i) => trait_ = i.0,
+        };
+
+        for m in trait_.module(db).path_to_root(db).iter().rev() {
+            if let Some(name) = m.name(db) {
+                s.push_str(&format!("{}::", name.to_smol_str()));
+            } else {
+                continue;
+            }
+        }
+
+        s.push_str(&trait_.name(db).to_smol_str());
+        s
+    }
+}
+
+/// A utility struct that is used for the enclosing struct.
+struct Struct {
+    strukt: ast::Struct,
+    name: ast::Name,
+}
+
+impl Struct {
+    pub(crate) fn new(s: ast::Struct) -> Option<Self> {
+        let name = s.name()?;
+        Some(Struct { name, strukt: s })
+    }
+
+    pub(crate) fn delegate(&self, field: Field, acc: &mut Assists, ctx: &AssistContext<'_>) {
+        let db = ctx.db();
+        for delegee in &field.impls {
+            // FIXME :  We can omit already implemented impl_traits
+            // But we don't know what the &[hir::Type] argument should look like.
```

Comment:
For this I think we need to make use of `TraitRef`s instead of plain `Trait`s. `TraitRef`s basically consist of a `Trait` and it's arguments, given we expose `TraitRef`s in hir now we might as well want to change `Type::impls_trait` to take a TraitRef instead, or make a new method that accepts one. But let's leave that for a follow up